feat: FAQ 페이지 비동기 처리 및 예외 처리

newminkyung opened the pull request:
## 이슈 번호

<!-- 관련 이슈 번호를 연결 -->

- resolve: #18 

## 구현 기능

### 1. useSuspenseQuery
- `useQuery`의 반환값과 대부분 동일하나 data가 non-nullable
  - useQuery가 반환하는 data는 `undefined | TData`

``` tsx
const { data } = useSuspendedQuery(['key'], queryFn);
```

### 2. Accordion 데이터가 없을 때 예외 처리
- `등록된 내용이 없습니다` 표시
- 아래 스크린샷 참고

## 변경 로직

- useFaqQuery: `useQuery` → `useSuspenseQuery`
  - data에 `undefined`가 반환되지 않음
  - nullable 데이터 관련 처리 삭제 (옵셔널 체이닝, null 병합 연산자 삭제)

- react-query `{ suspense: true }` option 추가
  - React.Suspense를 사용하기 위한 옵션

## 스크린샷

<!--해당 PR에 대한 이미지 -->

<img width="600" alt="image" src="https://user-images.githubusercontent.com/80238096/215024149-9cef5b0d-69f8-4c4e-952c-465a165a9363.png">


## 참고 사항

<!-- 궁금한 점, 논의할 점 등 -->

- 비동기 처리 관련 내용은 이슈(#18 ) 또는 [노션에 정리해놓은 글](https://www.notion.so/sejong-sos/Suspense-92feaaa6ea204f5b9308c99d224ba9e6)을 참고해주세요

- 예외 메세지 내용 확인 해주세요!
  - 내용 정해지면 나중에 constant로 분리해서 통일시키면 좋을 것 같습니다!
- fallback UI에 대해 추가 논의 필요

newminkyung committed:
feat: react-query suspense 옵션 설정 (#18)

- useSuspenseQuery: data가 non-nullable, 나머지는 react-query와 동일

Resolve: #18

newminkyung committed:
refactor: FAQ nullable 데이터 처리 삭제

- useSuspenseQuery를 사용하면 non-nullable 데이터만 반환받기 때문에 nullable 데이터 처리 삭제
  - 옵셔널 체이닝, null 병합 연산자 삭제

Related to: #18

newminkyung committed:
refactor: Accordion details가 빈 배열일 때 예외 처리

inyeong-kang commented on src/components/molecule/Accordion.tsx:
```diff
@@ -13,12 +13,16 @@ type AccordionItemProps<T extends React.ElementType> = Component<T>;
+          </details>
+        ))
+      ) : (
+        <ErrorMessage className="text-disabled-400 text-base">등록된 내용이 없습니다.</ErrorMessage>
```
ErrorMessage 요거는 공지사항이랑 건의사항 등등에도 재사용 가능해보이네요! 👍👍

inyeong-kang replied:
사용자가 데이터를 아직 등록하지 않은 경우, 예를 들어 전체 문제에 어떤 문제도 등록하지 않았을 때는 Error 대신 다른 용어를 써야할 듯 합니다~~

inyeong-kang replied:
정리해주신 노션 잘 읽었습니다.
그동안 전 명령형 컴포넌트 방식만 이용해서 로딩이랑 에러 처리까지 했었는데😅, Suspense Query를 쓰면 예외 처리가 좀 더 간결해지는 느낌이네요! (결국 사용자는 데이터를 성공적으로 fetch한(로딩 완료된) 상태의 컴포넌트에 집중하게 되니까..)
고생하셨습니다👍

newminkyung replied:
HTTP Status에서 에러가 전달된 경우만 에러로 취급하고 데이터가 없는 경우 등은 따로 생각해봐야겠네요! 
이 부분은 논의를 통해서 수정하기로 하고 우선 주석으로 남겨놓도록 하겠습니다!

newminkyung committed:
refactor: Accordion details가 빈 배열일 때 FIXME 주석 추가

- ErrorMessage가 아니라 다른 용어를 사용하기로 논의
- 어떤 용어를 사용할 지는 추후 논의를 통해 수정하기로 함

Related to: #19
